squidpy/pl/_spatial_utils.py

```diff
@@ -0,0 +1,897 @@
+        if color_vector.isna().any():
+            color_vector = color_vector.add_categories([to_hex(na_color)])
+            color_vector = color_vector.fillna(to_hex(na_color))
+        return color_source_vector, color_vector, True
+
+
+def _shaped_scatter(
+    x: NDArrayA,
+    y: NDArrayA,
+    s: float,
+    c: NDArrayA,
+    shape: _AvailShapes | None = ScatterShape.CIRCLE.s,  # type: ignore[assignment]
+    norm: _Normalize = None,
+    **kwargs: Any,
+) -> PatchCollection:
+    """
+    Get shapes for scatterplot.
+
+    Adapted from here: https://gist.github.com/syrte/592a062c562cd2a98a83 .
+    This code is under [The BSD 3-Clause License](http://opensource.org/licenses/BSD-3-Clause)
+    """
+    shape = ScatterShape(shape)  # type: ignore[assignment]
+    if TYPE_CHECKING:
+        assert isinstance(shape, ScatterShape)
+    shape = shape.s
+
+    if shape == ScatterShape.CIRCLE.s:
+        zipped = np.broadcast(x, y, s)
+        patches = [Circle((x_, y_), s_) for x_, y_, s_ in zipped]
+    else:
+        n = 4 if shape == ScatterShape.SQUARE.s else 6
+        r: float = s / (2 * np.sin(np.pi / n))
+        polys = np.stack([_make_poly(x, y, r, n, i) for i in range(n)], 1).swapaxes(0, 2)
+        patches = [Polygon(p, False) for p in polys]
+    collection = PatchCollection(patches, **kwargs)
+
+    if isinstance(c, np.ndarray) and np.issubdtype(c.dtype, np.number):
+        collection.set_array(np.ma.masked_invalid(c))
+        collection.set_norm(norm)
+    else:
+        collection.set_facecolor(c)
+
+    return collection
+
+
+def _make_poly(x: NDArrayA, y: NDArrayA, r: float, n: int, i: int) -> Tuple[NDArrayA, NDArrayA]:
+    x_i = x + r * np.sin((np.pi / n) * (1 + 2 * i))
+    y_i = y + r * np.cos((np.pi / n) * (1 + 2 * i))
+    return x_i, y_i
+
+
+def _plot_edges(
+    adata: AnnData,
+    coords: NDArrayA,
+    ax: Axes,
+    edges_width: float = 0.1,
+    edges_color: str | Sequence[float] | Sequence[str] = "grey",
+    connectivity_key: str = Key.obsp.spatial_conn(),
+    edges_kwargs: Mapping[str, Any] = MappingProxyType({}),
+) -> Any:
+    """Graph plotting."""
+    from networkx import Graph
+    from networkx.drawing.nx_pylab import draw_networkx_edges
+
+    if connectivity_key not in adata.obsp:
+        raise KeyError(
+            f"Unable to find `connectivity_key: {connectivity_key}` in `adata.obsp`.\
+             Please set `connectivity_key`."
+        )
+
+    g = Graph(adata.obsp[connectivity_key])
+    edge_collection = draw_networkx_edges(
+        g, coords, width=edges_width, edge_color=edges_color, arrows=False, ax=ax, **edges_kwargs
+    )
+    edge_collection.set_rasterized(sc_settings._vector_friendly)
+
+    return edge_collection
+
+
+def _get_title_axlabels(
+    title: _SeqStr, axis_label: _SeqStr, spatial_key: str, n_plots: int
+) -> Tuple[_SeqStr, Sequence[str]]:
+
+    # handle title
+    if title is not None:
+        if isinstance(title, list) and len(title) != n_plots:
```

Comment:
Would use `isinstance(title, (tuple, list))` just in case, but no strong opinion.